spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/actions/DeleteOrphanFilesSparkAction.java

```diff
@@ -303,6 +304,19 @@ private Dataset<String> listedFileDS() {
     List<String> subDirs = Lists.newArrayList();
     List<String> matchingFiles = Lists.newArrayList();
 
+    if (table.io() instanceof SupportsPrefixOperations) {
+      Iterator<org.apache.iceberg.io.FileInfo> iterator = ((SupportsPrefixOperations) table.io()).listPrefix(location).iterator();
```

Comment:
I think this is a good start but I would ask that we only make the change in v3.4 to start (we can backport level)

In addition can we set this up without the early exit?

Ie
```java
If (table.io supportsPrefix)
 files = listPrefix()
else
  files = listNoPrefix()
  }
  
  spark.createDataset(files)
  ```
  
  We also need a test which exercises this code path, (Does HadoopFS do this by default? If So do we have a test for the other path:noprefix)

Reply:
Thanks @RussellSpitzer for taking a look. I can make the change to 3.4 only.
However I think early exit would be ideal, as after this `if` condition we would be calling the

```
    listDirRecursively(
        location,
        predicate,
        hadoopConf.value(),
        MAX_DRIVER_LISTING_DEPTH,
        MAX_DRIVER_LISTING_DIRECT_SUB_DIRS,
        subDirs,
        pathFilter,
        matchingFiles);

.......
 ListDirsRecursively listDirs = new ListDirsRecursively(conf, olderThanTimestamp, pathFilter);
 JavaRDD<String> matchingLeafFileRDD = subDirRDD.mapPartitions(listDirs);
```
Which would both be making direct FileSystem `listStatus` calls which we ideally want to avoid and just use the S3fileIO. Let me know if there are concerns with the early exit/current logic 

cc @jackye1995

Reply:
The idea is not to change the logic, but to change the structure of the code. Basically you would take this code and the code after the exit and create two functions (names are just examples)


```java
Dataset<String> listWithPrefix() {
    List<String> matchingFiles = Lists.newArrayList();
    Iterator<org.apache.iceberg.io.FileInfo> iterator = ((SupportsPrefixOperations) 
    table.io()).listPrefix(location).iterator();
    while (iterator.hasNext()) {
      org.apache.iceberg.io.FileInfo fileInfo = iterator.next();
      if (fileInfo.createdAtMillis() < olderThanTimestamp) {
          matchingFiles.add(fileInfo.location());
      }
    }
    JavaRDD<String> matchingFileRDD = sparkContext().parallelize(matchingFiles, 1);
    return spark().createDataset(matchingFileRDD.rdd(), Encoders.STRING());
  }

Dataset<String> listWithoutPrefix() {
    List<String> matchingFiles = Lists.newArrayList();
    Predicate<FileStatus> predicate = file -> file.getModificationTime() < olderThanTimestamp;
    PathFilter pathFilter = PartitionAwareHiddenPathFilter.forSpecs(table.specs());

    // list at most MAX_DRIVER_LISTING_DEPTH levels and only dirs that have
    // less than MAX_DRIVER_LISTING_DIRECT_SUB_DIRS direct sub dirs on the driver
    listDirRecursively(
        location,
        predicate,
        hadoopConf.value(),
        MAX_DRIVER_LISTING_DEPTH,
        MAX_DRIVER_LISTING_DIRECT_SUB_DIRS,
        subDirs,
        pathFilter,
        matchingFiles);
        
    JavaRDD<String> matchingFileRDD = sparkContext().parallelize(matchingFiles, 1);
    if (subDirs.isEmpty()) {
      return spark().createDataset(matchingFileRDD.rdd(), Encoders.STRING());
    }
    int parallelism = Math.min(subDirs.size(), listingParallelism);
    JavaRDD<String> subDirRDD = sparkContext().parallelize(subDirs, parallelism);
    Broadcast<SerializableConfiguration> conf = sparkContext().broadcast(hadoopConf);
    ListDirsRecursively listDirs = new ListDirsRecursively(conf, olderThanTimestamp, pathFilter);
    JavaRDD<String> matchingLeafFileRDD = subDirRDD.mapPartitions(listDirs);
    JavaRDD<String> completeMatchingFileRDD = matchingFileRDD.union(matchingLeafFileRDD);
    return spark().createDataset(completeMatchingFileRDD.rdd(), Encoders.STRING());
}


Then you change listedFilesDS

listedFileDS() { 
If (table.io supportsPrefix)
 files = listPrefix()
else
  files = listNoPrefix()
  }
  return files
}
```

Reply:
I see, thank you for the detailed explanation

Reply:
Example of the pattern in ExpireSnapshots,
https://github.com/apache/iceberg/blob/master/spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/actions/ExpireSnapshotsSparkAction.java#L242-L269